pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package exponentialbackoff contains logic for implementing exponential
+// backoff for GoRoutineMap and NestedPendingOperations.
+package exponentialbackoff
+
+import (
+	"fmt"
+	"time"
+)
+
+const (
+	// initialDurationBeforeRetry is the amount of time after an error occurs
+	// that GoroutineMap will refuse to allow another operation to start with
+	// the same target (if exponentialBackOffOnError is enabled). Each
+	// successive error results in a wait 2x times the previous.
```

Comment:
> if the volume becomes attachable in 1 min, depending on the timing, it is possible for all N pods waiting for more than 3 mins before starting mount operations

That's not correct. If multiple pods on a node are referencing the same attachable volume, the attach and mount device steps are serialized and only happen *once* for all the pods. Once a device is attached and globally mounted, subsequent pods only need to do the bind mount (which, with this change is parallelized). 

> Also it may have higher chances of operation is rejected because there is an existing operation.

That makes little difference in execution speed. The reconciler retries very quickly.

> What will happen if there is a hanging operation in goroutine

If an attachable volume fails to attach or complete global mount, regardless of if the operations fail or hang, they will block bind mounting. That is necessary and by design, an attachable volume can not be bind mounted until these operations succeed.

> For volumes that are not attachable, since they can run in parallel without problem. Why need to run them through goroutine?

They must be run in a goroutine so that the caller is not blocked. They are run using `nestedpendingoperations` to prevent the multiple operations from being triggered on the same pod/volume.